crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3655,7 +3655,10 @@ impl<'db> TypeInferenceBuilder<'db> {
             // TODO: parse the expression and check whether it is a string annotation, since they
             // can be annotation expressions distinct from type expressions.
             // https://typing.readthedocs.io/en/latest/spec/annotations.html#string-annotations
-            ast::Expr::StringLiteral(_literal) => Type::Todo,
+            ast::Expr::StringLiteral(_literal) => {
+                self.store_expression_type(expression, Type::Todo);
+                Type::Todo
+            }
 
             // Annotation expressions also get special handling for `*args` and `**kwargs`.
             ast::Expr::Starred(starred) => self.infer_starred_expression(starred),
```

Comment:
I suspect we have the same bug here, for starred expressions in an annotation.

In general the contract is supposed to be that `infer_expression` takes care of storing the expression type, so we don't have to do it separately for every kind of expression.

For annotations, the tricky thing is the split between annotation expressions (what is valid in an annotation) and type expressions (which is a subset of annotation expressions). Currently `infer_type_expression` stores the type, but `infer_annotation_expression` does not, because most of the time it just defers to `infer_type_expression`, and we don't want to double-store. But we should come up with some general way to ensure annotation expression types are always stored, without having to do it separately in each branch that we add here.

Reply:
Oh, I *did* look at `infer_starred_expression` and saw that it uses `infer_expression` (which stores the expression type), but that's probably not enough… I'll look into it!

Reply:
It uses `infer_expression` on a sub-part of the starred expression (the `value`), but never on the starred expression itself. This is the usual pattern for `infer_*` methods for particular node types, since the assumption is they are called from `infer_expression` which will store the type they return.